### PR TITLE
Adding maxPaths option to Router.

### PR DESCRIPTION
**Pull request by jhusain**
# Limiting the number of paths returned by the Router

Prior to version 0.4.0, the Falcor Router had a serious DOS vulnerability: requesting a large range of values could pin the CPU of the server running the Router.

Consider the following example in which the name of huge number of genrelists is requested from the Router:

```js
// falcor-router v0.3.0 and below
var Router = require('falcor-router');
var router = new Router(
  [
    {
      route: "genrelists[{integers}].name",
      get: function(pathSet) {
        // code to return the names of each of the requested genre lists
      }
    }
  ]);

// pins CPU!
router.
  get(["genrelists", {from:0, to: Number.MAX_SAFE_INTEGER}, "name"]).
  subscribe({ 
    onNext: function(jsonGraphEnvelope) { console.log(JSON.stringify(jsonGraphEnvelope)); },
    onError: function(error) { console.error(error); },
    onCompleted: function() { console.log("done"); })
```

## The `maxPaths` option

The DOS vulnerability has been mitigated in v0.4.0 by adding a `maxPaths` value to the options object accepted by the Router constructor. This value specifies the maximum number of paths which can be returned by a Falcor Router operation (either get, set, or call). The `maxPaths` value has a somewhat arbitrary default of 9000, which should provide some basic protection against a DOS.

```js
// falcor-router v0.4.0 and above
var Router = require('falcor-router');
var MaxPathsExceeded = require('falcor-router/errors/MaxPathsExceeded');
var router = new Router(
  [
    {
      route: "genrelists[{integers}].name",
      get: function(pathSet) {
        // code to return the names of each of the requested genre lists
      }
    }
  ], { maxPaths: 500 });

// prints "You asked for too many paths."
router.
  get(["genrelists", {from:0, to: Number.MAX_SAFE_INTEGER}, "name"]).
  subscribe({ 
    onNext: function(jsonGraphEnvelope) { console.log(JSON.stringify(jsonGraphEnvelope)); },
    onError: function(error) { 
      if (error instanceof MaxPathsExceeded) {
        console.error("You asked for too many paths."); 
      }
    },
    onCompleted: function() { console.log("done"); })
```

## Setting maximum path limits for individual routes

The `maxPaths` value allows you to set a global limit on the number of paths returned by a Router operation. However developers may want to set different maximums for different routes. For example you may want to allow no more than 10 titles and no more than 100 people to be retrieved in a single operation. To accomplish this, developers can test for the size of requested paths manually within a route handler and throw the `MaxPathsExceeded` error if too many paths are requested.

```js
// falcor-router v0.4.0 and above
var Router = require('falcor-router');
var MaxPathsExceeded = require('falcor-router/errors/MaxPathsExceeded');
var router = new Router(
  [
    {
      route: "genrelists[{integers}].name",
      get: function(pathSet) {
        // limit the number of requested genrelists to 200
        if (pathSet[1].length > 200) {
          throw new MaxPathsExceededError("You exceeded the maximum number of genre lists that can be requested in a single operation.");
        }
        // code to return the names of each of the requested genre lists
      }
    }
  ], { maxPaths: 500 });

// prints "You exceeded the maximum number of genre lists that can be requested in a single operation."
router.
  get(["genrelists", {from:0, to: Number.MAX_SAFE_INTEGER}, "name"]).
  subscribe({ 
    onNext: function(jsonGraphEnvelope) { console.log(JSON.stringify(jsonGraphEnvelope)); },
    onError: function(error) { 
      if (error instanceof MaxPathsExceeded) {
        console.error(error.message); 
      }
    },
    onCompleted: function() { console.log("done"); })
```

### CODE REVIEW

**Comment by jhusain** [comment]
FYI @sdesai

File: src/router/set.js [hunk 1]
```diff
@@ -13,131 +13,144 @@ var normalizePathSets = require('../operations/ranges/normalizePathSets');
...
-                    pathIntersection.
-                        reduce(function(acc, pathValue) {
-                            pathValueMerge(acc, pathValue);
+
+    return Observable.defer(function() {
+        var jsongCache = {};
+        var action = runSetAction(router, jsonGraph, jsongCache);
+        jsonGraph.paths = normalizePathSets(jsonGraph.paths);
+
+        if (getPathsCount(jsonGraph.paths) > router.maxPaths) {
```
Comment (sdesai) [inline src/router/set.js]: Same here? I imagine this is the exact same set impl as before just with the path check, on subscription?

File: src/router/call.js [hunk 2]
```diff
@@ -13,54 +16,65 @@ var collapse = pathUtils.collapse;
...
+    return Observable.defer(function() {
+
+        var refPaths = normalizePathSets(refPathsArg || []);
+        var thisPaths = normalizePathSets(thisPathsArg || []);
+        var jsongCache = {};
+        var action = runCallAction(router, callPath, args,
+                                   refPaths, thisPaths, jsongCache);
+        var callPaths = [callPath];
+
+        if (getPathsCount(refPaths) +
```
Comment (sdesai) [inline src/router/call.js]: Same here. I imagine this is the exact same call impl as before just with the path count check, on subscription?

**Comment by sdesai** [comment]
Reviewed. Looks good to me. Couple of comments inline, mainly just because the diffs are not as clear to read as they could be.

File: src/router/get.js [hunk 3]
```diff
@@ -5,55 +5,68 @@ var normalizePathSets = require('../operations/ranges/normalizePathSets');
...
-            return materialize(router, normPS, jsonGraphEnvelope);
-        });
+
+    return Observable.defer(function() {
+
+        var jsongCache = {};
+        var action = runGetAction(router, jsongCache);
+        var normPS = normalizePathSets(paths);
+
+        if (getPathsCount(normPS) > router.maxPaths) {
```
Comment (sdesai) [inline src/router/get.js]: It's tough to tell from the diff, but I assume this is just the exact same impl as before, but just checks for getPathsCount on any subscription?

If so, this looks fine.

**Comment by tivac** [comment]
@jhusain I know this already landed, but why don't individual routes support the `maxPaths : <integer>` option? 

I assume I'm missing some hidden complexity, but that seems like a nice way to reduce the amount of boilerplate required if you have lots of routes that want to set their own limits on the number of paths returned.

**Comment by jhusain** [comment]
I think that could be some nice sugar. We'll probably add that in the future. Note that I want to preserve the ability to throw programmatically. One reason is that you might want to set different rate limits depending on the operation (set vs. get vs. call). For example, you might want to only allow setting 10 videos, but getting 100. Given the potential differences in cost for writes vs. reads this seems like it could be useful.

**Comment by tivac** [comment]
Ah, sure. Something like

```js
// Simple case, applies to all operations
maxPaths : 10,

// Specify defaults per-operation
maxPaths : {
    get  : 100,
    set  : 10,
    call : 1
}
```

might be nice there too (apologies for the bikeshedding!)

I've got zero issues w/ the ability to throw programmatically, having it be part of the route definition seems like an easy win for reducing the complexity of route-handlers by taking a common-ish thing and making it not manual.

**Comment by jhusain** [comment]
I like your proposal.

**Comment by mik01aj** [comment]
And what would happen when the router gets a request for more routes then some handler can handle? Would the router look for other matching routes (with lower priority), or will it fail that subtree (with empty atoms, or errors at leaves), or fail the entire request?

**Comment by jhusain** [comment]
This error is fatal and causes the request to fail. There is no fallback. It is considered a bad request error (400 series).

**Pull request merged**